File: behavioral/dart/template_method_structured/report_generator.dart
abstract class ReportGenerator {
  void generateReport() {
    print('\n=== Generating Report ===');
    collectData();
    analyzeData();
    formatReport();
    exportReport();
    print('=== Report Generated ===');
  }
  
  void collectData();
  void analyzeData();
  void formatReport();
  void exportReport();
}

class SalesReport extends ReportGenerator {
  @override
  void collectData() {
    print('Collecting sales data from database...');
  }
  
  @override
  void analyzeData() {
    print('Analyzing sales trends...');
    print('Sales increased by 15% this quarter');
  }
  
  @override
  void formatReport() {
    print('Formatting sales report with charts...');
  }
  
  @override
  void exportReport() {
    print('Exporting to PDF...');
  }
}

class InventoryReport extends ReportGenerator {
  @override
  void collectData() {
    print('Collecting inventory data...');
  }
  
  @override
  void analyzeData() {
    print('Analyzing stock levels...');
    print('3 items need restocking');
  }
  
  @override
  void formatReport() {
    print('Formatting inventory report with tables...');
  }
  
  @override
  void exportReport() {
    print('Exporting to Excel...');
  }
}

File: behavioral/typescript/template_method_structured/report_generator.ts
export abstract class ReportGenerator {
  generateReport(): void {
    console.log('\n=== Generating Report ===');
    this.collectData();
    this.analyzeData();
    this.formatReport();
    this.exportReport();
    console.log('=== Report Generated ===');
  }
  
  protected abstract collectData(): void;
  protected abstract analyzeData(): void;
  protected abstract formatReport(): void;
  protected abstract exportReport(): void;
}

export class SalesReport extends ReportGenerator {
  protected collectData(): void {
    console.log('Collecting sales data from database...');
  }
  
  protected analyzeData(): void {
    console.log('Analyzing sales trends...');
    console.log('Sales increased by 15% this quarter');
  }
  
  protected formatReport(): void {
    console.log('Formatting sales report with charts...');
  }
  
  protected exportReport(): void {
    console.log('Exporting to PDF...');
  }
}

export class InventoryReport extends ReportGenerator {
  protected collectData(): void {
    console.log('Collecting inventory data...');
  }
  
  protected analyzeData(): void {
    console.log('Analyzing stock levels...');
    console.log('3 items need restocking');
  }
  
  protected formatReport(): void {
    console.log('Formatting inventory report with tables...');
  }
  
  protected exportReport(): void {
    console.log('Exporting to Excel...');
  }
}
